src/vm/gcheaputilities.cpp

```diff
@@ -80,6 +80,105 @@ HMODULE GCHeapUtilities::GetGCModule()
 namespace
 {
 
+// This block of code contains all of the state necessary to handle incoming
+// EtwCallbacks before the GC has been initialized. This is a tricky problem
+// because EtwCallbacks can appear at any time, even when we are just about
+// finished initializing the GC.
+//
+// The below lock is taken by the "main" thread (the thread in EEStartup) and
+// the "ETW" thread, the one calling EtwCallback. EtwCallback may or may not
+// be called on the main thread.
+CRITICAL_SECTION g_eventStashLock;
+
+// These two global fields are 64-bit integers so that loads and stores to them
+// can be atomic.
+LONG64 g_stashedKeywordAndLevel = 0;
+LONG64 g_stashedPrivateKeywordAndLevel = 0;
+
+// In order for loads and stores of the keyword and level state to be atomic,
+// these macros assist in packing and unpacking two 32-bit integer values into
+// a 64-bit value.
+#define STASH_KEY_AND_LEVEL(key, level) ((static_cast<LONG64>(key) << 32) | static_cast<LONG64>(level))
+#define UNSTASH_KEY(combined) (static_cast<GCEventKeyword>(combined >> 32))
+#define UNSTASH_VALUE(combined) (static_cast<GCEventLevel>(combined & 0xFFFFFFFF))
+
+// FinalizeLoad is called by the main thread to complete initialization of the GC.
+// At this point, the GC has provided us with an IGCHeap instance and we are preparing
+// to "publish" it by assigning it to g_pGCHeap.
+//
+// This function can proceed concurrently with StashKeywordAndLevel below.
+void FinalizeLoad(IGCHeap* gcHeap, IGCHandleManager* handleMgr, HMODULE gcModule)
+{
+    // at this point the GC has been initialized and is ready to go.
+    // the main thread is responsible for initializing the lock.
+    UnsafeInitializeCriticalSection(&g_eventStashLock);
+    UnsafeEnterCriticalSection(&g_eventStashLock);
+
+    // the compiler or hardware can't be allowed to reorder this write before
+    // the lock acquision. This is because StashKeywordAndLevel will acquire the
+    // lock if g_pGCHeap is not null and the lock must be initialized by that point.
+    VOLATILE_MEMORY_BARRIER();
```

Comment:
This is not necessary. Locks imply memory barrier.

Reply:
ok!